### ISSUE TITLE
NullPointerException on RealmChangeListener with async queries workflow

### ISSUE THREAD

**Issue by jpmcosta** [realm/realm-java#4495]
#### Goal

By making use of two `RealmChangeListeners` and one async query, my goal is to reutilize objects when possible, or create a new ones if needed.

#### Results

- Before Realm `3.0` this workflow worked well;
- On `3.0.0` it stopped working, because of https://github.com/realm/realm-java/issues/4352;
- On `3.1.1` it throws a `NullPointerException`.

```
java.lang.NullPointerException: Attempt to invoke virtual method 'void io.realm.internal.Collection.removeListener(java.lang.Object, io.realm.RealmChangeListener)' on a null object reference
	at io.realm.internal.PendingRow.clearPendingCollection(PendingRow.java:205)
	at io.realm.internal.PendingRow.notifyFrontEnd(PendingRow.java:236)
	at io.realm.internal.PendingRow.access$000(PendingRow.java:18)
	at io.realm.internal.PendingRow$1.onChange(PendingRow.java:46)
	at io.realm.internal.PendingRow$1.onChange(PendingRow.java:43)
	at io.realm.internal.Collection$RealmChangeListenerWrapper.onChange(Collection.java:62)
	at io.realm.internal.Collection$CollectionObserverPair.onChange(Collection.java:43)
	at io.realm.internal.Collection$Callback.onCalled(Collection.java:87)
	at io.realm.internal.Collection$Callback.onCalled(Collection.java:77)
	at io.realm.internal.ObserverPairList.foreach(ObserverPairList.java:108)
	at io.realm.internal.Collection.notifyChangeListeners(Collection.java:524)
	at android.os.MessageQueue.nativePollOnce(Native Method)
	at android.os.MessageQueue.next(MessageQueue.java:143)
	at android.os.Looper.loop(Looper.java:122)
	at android.app.ActivityThread.main(ActivityThread.java:5254)
	at java.lang.reflect.Method.invoke(Native Method)
	at java.lang.reflect.Method.invoke(Method.java:372)
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:903)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:698)
```

#### Steps & Code to Reproduce

- The bug seems to occur when the last `Transaction` tries to call its listeners, after `item.reset();`
- `doThings(item);` is never called.

##### Classes

```java
public class Item extends RealmObject {

    public static Item create(Long id) {
        Item item = new Item();
        item.id = id;
        return item;
    }

    @PrimaryKey
    public Long id;

    public String name;

    public Feed parentFeed;


    public void reset() {
        name = "";
    }
}
```
```java
public class Feed extends RealmObject {

    @PrimaryKey
    public Long id;

    public String name;
}
```

##### Code Sample

```java
final Item item = getRealm().where(Item.class).isNull("parentFeed").findFirstAsync();

// Create a new item, if no valid item is found.
item.addChangeListener(new RealmChangeListener<Item>() {

    @Override
    public void onChange(Item item) {
        if (!item.isValid()) {
            item.removeChangeListener(this);

            getRealm().executeTransaction(new Realm.Transaction() {

                @Override
                public void execute(Realm realm) {
                    getRealm().copyToRealm(Item.create(getLastId()));
                }
            });
        }
    }
});

// Wait for valid item, reset, and use it to do things.
item.addChangeListener(new RealmChangeListener<Item>() {

    @Override
    public void onChange(final Item item) {
        if (item.isValid()) {
            item.removeChangeListener(this);

            getRealm().executeTransaction(new Realm.Transaction() {

                @Override
                public void execute(Realm realm) {
                    item.reset();
                }
            });

            doThings(item);
        }
    }
});
```

#### Notes

I could use just a single `RealmChangeListener`, but I'm using Kotlin in my project, and I'm making use of two different extensions functions, which is why I end up using two different listeners.

#### Version of Realm

Realm version(s): 3.1.1

**Comment by Zhuinden** [comment]
Is `Item` retained as a field reference?

If it's a local reference, then it can get nulled out at pretty much any time.

**Comment by jpmcosta** [comment]
Yes, even though the code sample doesn't reflect that, `Item` is a field reference.
Thank you, for allowing me to clarify that. 👍

**Comment by beeender** [comment]
@jpmcosta Thanks a lot for finding another bug! This will be fixed in #4497 

But there is still a problem with you implementation, from `v3.1.0` we fixed an inconsistency behaviour of `findFirstAsync()` 
>  `findFirstAsync()` now returns an invalid object if there is no object matches the query condition instead of running the query repeatedly until it can find one (#4352).

That means in your case, if there is no object match the query, your second listener won't be triggered since the `item` will be an invalid object forever unless you re-query for it.

Maybe that is your expected behavior? 

If you still want to use two listeners before the fix released, you can just change the `removeChangeListener` to `removeAllChangeListeners()` in the first listener? Since anyway in this case, the 2nd listener won't do anything meaningful.

**Comment by jpmcosta** [comment]
I see. I thought it would receive the invalid object, but it would still be listening for changes for the same query.
I'll probably have to update my workflow. Thanks!
